RecoParticleFlow/PFSimProducer/plugins/EcalBarrelClusterFastTimer.cc

```diff
@@ -0,0 +1,220 @@
+    filler.fill();
+    iEvent.put(valMap, label);
+  }
+}
+
+EcalBarrelClusterFastTimer::EcalBarrelClusterFastTimer(const edm::ParameterSet& conf) :
+  ebTimeHitsToken_(consumes<EcalRecHitCollection>( conf.getParameter<edm::InputTag>("ebTimeHits") ) ),
+  ebClustersToken_(consumes<std::vector<reco::PFCluster> >( conf.getParameter<edm::InputTag>("ebClusters") ) ),
+  timedVertexToken_(consumes<reco::VertexCollection>( conf.getParameter<edm::InputTag>("timedVertices") ) ),
+  minFraction_( conf.getParameter<double>("minFractionToConsider") ),
+  minEnergy_(conf.getParameter<double>("minEnergyToConsider") ),
+  ecalDepth_(conf.getParameter<double>("ecalDepth") )
+{
+  // setup resolution models
+  const std::vector<edm::ParameterSet>& resos = conf.getParameterSetVector("resolutionModels");
+  for( const auto& reso : resos ) {
+    const std::string& name = reso.getParameter<std::string>("modelName");
+    ResolutionModel* resomod = ResolutionModelFactory::get()->create(name,reso);
+    _resolutions.emplace_back( resomod );  
+
+    // times and time resolutions for general tracks
+    produces<std::vector<std::vector<float> > >(name); // indexed by primary vertex then PF cluster
+    produces<edm::ValueMap<float> >(name+resolution);    
+  }
+  // get RNG engine
+  edm::Service<edm::RandomNumberGenerator> rng;
+  if (!rng.isAvailable()){
+    throw cms::Exception("Configuration")
+      << "EcalBarrelClusterFastTimer::EcalBarrelClusterFastTimer() - RandomNumberGeneratorService is not present in configuration file.\n"
+      << "Add the service in the configuration file or remove the modules that require it.";
+  }
+  _rng_engine = &(rng->getEngine());
+}
+
+void EcalBarrelClusterFastTimer::produce(edm::Event& evt, const edm::EventSetup& es) {
+  edm::Handle<std::vector<reco::PFCluster> > clustersH;
+  edm::Handle<EcalRecHitCollection> timehitsH;
+  edm::Handle<reco::VertexCollection> verticesH;
+
+  evt.getByToken(ebClustersToken_,clustersH);
+  evt.getByToken(ebTimeHitsToken_,timehitsH);
+  evt.getByToken(timedVertexToken_,verticesH);
+  
+   edm::ESHandle<CaloGeometry> geoHandle;
+   es.get<CaloGeometryRecord>().get(geoHandle);
+   const auto* barrelGeom = 
+     geoHandle->getSubdetectorGeometry(DetId::Ecal,EcalBarrel);
+
+  const auto& clusters = *clustersH;
+  const auto& timehits = *timehitsH;
+  const auto& vertices = *verticesH;
+
+  std::vector<std::pair<float,DetId> > times; // perfect times keyed to cluster index
+  times.reserve(clusters.size());
+  
+  for( const auto& cluster : clusters ) {
+    times.push_back( getTimeForECALPFCluster( timehits, cluster ) );
+  }
+  
+  for( const auto& reso : _resolutions ) {
+    const std::string& name = reso->name();
+    std::vector<float> resolutions;
+    std::vector<std::pair<float,DetId> > smeared_times;
+    resolutions.reserve(clusters.size());
+    smeared_times.reserve(clusters.size());
+    std::auto_ptr<std::vector< std::vector<float> > > outP( new std::vector< std::vector<float> > );
```

Comment:
Would it be possible to address this in the next PR where this code is introduced into the reconstruction workflows? Unless there are other major comments.